src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs

```diff
@@ -109,10 +109,33 @@ public async Task<IEnumerable<ContentItem>> GetAsync(IEnumerable<string> content
             }
             else
             {
-                contentItems = (await _session
-                    .Query<ContentItem, ContentItemIndex>()
-                    .Where(x => x.ContentItemId.IsIn(contentItemIds) && x.Published == true)
-                    .ListAsync()).ToList();
+                contentItems = new List<ContentItem>();
+                var contentItemIdsAdded = new List<string>();
+
+                foreach (var contentItemId in contentItemIds)
+                {
+                    // If the published version is already loaded, we can return it.
+                    if (_contentManagerSession.RecallPublishedItemId(contentItemId, out var contentItem))
+                    {
+                        contentItems.Add(contentItem);
+                        contentItemIdsAdded.Add(contentItemId);
+                    }
+                }
+
+                // Query only the ids not already added
+                var contentItemIdsToQuery = contentItemIds.Except(contentItemIdsAdded);
+
+                // Get the remaining content items
+                if (contentItemIdsToQuery.Any())
+                {
+                    var contentItemsToLoad = (await _session
+                           .Query<ContentItem, ContentItemIndex>()
+                           .Where(x => x.ContentItemId.IsIn(contentItemIdsToQuery) && x.Published == true)
+                           .ListAsync()).ToList();
+
+                    contentItems.AddRange(contentItemsToLoad);
+                    contentItems.ForEach(c => _contentManagerSession.Store(c));                    
```

Comment:
We don't need to `Store()` the items that we retrieved from the store, in fact we also don't need to store here the items that we just queried because **this is `LoadAsync()` that should do it** below and then call the related handlers.

Then line 143, normally we don't need to call `LoadAsync()` on the items that we retrieved from the store, but that's okay as `LoadAsync()` checks if it is already stored (so already queried and then loaded).

One suggestion would be something like this, not tried, quickly done ;)
Note: A little complicated at the end, but to prevent useless allocations

            ...
            ...
            List<ContentItem> contentItems = null;
            List<ContentItem> storedItems = null;

            if (latest)
            {
                contentItems = (await _session
                    .Query<ContentItem, ContentItemIndex>()
                    .Where(x => x.ContentItemId.IsIn(contentItemIds) && x.Latest == true)
                    .ListAsync()).ToList();
            }
            else
            {
                foreach (var contentItemId in contentItemIds)
                {
                    // If the published version is already stored, we can return it.
                    if (_contentManagerSession.RecallPublishedItemId(contentItemId, out var contentItem))
                    {
                        if (storedItems == null)
                        {
                            storedItems = new List<ContentItem>();
                        }

                        storedItems.Add(contentItem);
                    }
                }

                // Only query the ids not already stored.
                var itemIdsToQuery = storedItems != null
                    ? contentItemIds.Except(storedItems.Select(x => x.ContentItemId))
                    : contentItemIds;

                if (itemIdsToQuery.Any())
                {
                    contentItems = (await _session
                       .Query<ContentItem, ContentItemIndex>()
                       .Where(x => x.ContentItemId.IsIn(itemIdsToQuery) && x.Published == true)
                       .ListAsync()).ToList();
                }
            }

            if (contentItems != null)
            {
                for (var i = 0; i < contentItems.Count; i++)
                {
                    contentItems[i] = await LoadAsync(contentItems[i]);
                }

                if (storedItems != null)
                {

                    contentItems.AddRange(storedItems);
                }
            }
            else if (storedItems != null)
            {
                contentItems = storedItems;
            }
            else
            {
                return Enumerable.Empty<ContentItem>();
            }
            ...
            ...

Reply:
Ok @jtkech , I merged dev and did what you suggest.